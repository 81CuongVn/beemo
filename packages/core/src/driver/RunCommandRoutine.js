/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import { Routine } from 'boost';
import chalk from 'chalk';
import glob from 'glob';
import isGlob from 'is-glob';
import parseArgs from 'yargs-parser';

import type { DriverContext, Execution } from '../types';

type Args = string[];

const OPTION_PATTERN: RegExp = /-?-[-a-z0-9]+(,|\s)/gi;

export default class RunCommandRoutine extends Routine<Object, DriverContext> {
  execute(): Promise<*> {
    const { metadata } = this.context.primaryDriver;

    this.task('Gathering arguments', this.gatherArgs);

    this.task('Expanding glob patterns', this.expandGlobPatterns);

    this.task('Filtering options', this.filterUnknownOptions).skip(!metadata.filterOptions);

    this.task('Including config option', this.includeConfigOption).skip(!metadata.useConfigOption);

    this.task('Running command', this.runCommandWithArgs);

    return this.serializeTasks();
  }

  /**
   * Expand arguments that look like globs.
   */
  expandGlobPatterns(args: Args): Promise<Args> {
    const nextArgs = [];

    this.tool.debug('Expanding glob patterns');

    args.forEach(arg => {
      if (isGlob(arg)) {
        const paths = glob.sync(arg, {
          cwd: this.context.root,
          debug: this.tool.config.debug,
          strict: true,
        });

        this.tool.debug(`  ${arg} ${chalk.gray('->')} ${paths.join(', ')}`);

        nextArgs.push(...paths);
      } else {
        nextArgs.push(arg);
      }
    });

    return Promise.resolve(nextArgs);
  }

  /**
   * Extract native supported options and flags from driver help output.
   */
  extractNativeOptions(): Promise<{ [option: string]: true }> {
    const driver = this.context.primaryDriver;
    const { env } = driver.options;
    const options = driver.getSupportedOptions();

    if (options.length > 0) {
      this.tool.debug('Using supported options from driver');

      const nativeOptions = {};

      options.forEach(option => {
        nativeOptions[option] = true;
      });

      return Promise.resolve(nativeOptions);
    }

    this.tool.debug('Extracting native options from help output');

    return this.executeCommand(driver.metadata.bin, [driver.metadata.helpOption], { env }).then(
      ({ stdout }) => {
        const nativeOptions = {};

        stdout.match(OPTION_PATTERN).forEach(option => {
          // Trim trailing comma or space
          nativeOptions[option.slice(0, -1)] = true;
        });

        return nativeOptions;
      },
    );
  }

  /**
   * Filter unknown and or unsupported CLI options from the arguments passed to the CLI.
   * Utilize the driver's help option/command to determine accurate options.
   */
  filterUnknownOptions(args: Args): Promise<Args> {
    this.tool.debug('Filtering unknown command line options');

    return this.extractNativeOptions().then(nativeOptions => {
      const filteredArgs = [];
      const unknownArgs = [];
      let skipNext = false;

      args.forEach((arg, i) => {
        if (skipNext) {
          skipNext = false;

          return;
        }

        if (arg.startsWith('-')) {
          let option = arg;
          const nextArg = args[i + 1];

          // --opt=123
          if (option.includes('=')) {
            [option] = option.split('=');

            if (!nativeOptions[option]) {
              unknownArgs.push(arg);

              return;
            }

            // --opt 123
          } else if (!nativeOptions[option]) {
            unknownArgs.push(arg);

            if (nextArg && !nextArg.startsWith('-')) {
              skipNext = true;
              unknownArgs.push(nextArg);
            }

            return;
          }
        }

        filteredArgs.push(arg);
      });

      if (unknownArgs.length > 0) {
        this.tool.debug(`Filtered args: ${unknownArgs.join(', ')}`);
      }

      return filteredArgs;
    });
  }

  /**
   * Gather arguments from all sources to pass to the driver.
   */
  gatherArgs(): Promise<Args> {
    const driverArgs = this.context.primaryDriver.getArgs();
    const commandArgs = this.context.args;
    const args = [
      // Passed by the driver
      ...driverArgs,
      // Passed on the command line
      ...commandArgs,
      // Parallel args passed on the command line
      // ...(this.config.parallelArgs || []),
    ];

    this.tool.debug('Gathering arguments to pass to driver');

    this.tool.invariant(
      driverArgs.length > 0,
      '  From driver "args" option',
      driverArgs.join(' '),
      'No arguments',
    );

    this.tool.invariant(
      commandArgs.length > 0,
      '  From the command line',
      commandArgs.join(' '),
      'No arguments',
    );

    // Since we combine multiple args, we need to rebuild this.
    // And we also need to set this before we filter them.
    this.context.yargs = parseArgs(args);

    return Promise.resolve(args);
  }

  /**
   * Include --config option if driver requires it (instead of auto-lookup resolution).
   */
  includeConfigOption(prevArgs: Args): Promise<Args> {
    const { configPaths, primaryDriver } = this.context;
    const configPath = configPaths.find(path => path.endsWith(primaryDriver.metadata.configName));
    const args = [...prevArgs];

    if (configPath) {
      args.push(primaryDriver.metadata.configOption, configPath);
    }

    this.tool.debug('Including config option to args');

    return Promise.resolve(args);
  }

  /**
   * Execute the driver's command with the filtered arguments and handle the
   * success and failures with the driver itself.
   */
  runCommandWithArgs(args: Args): Promise<Execution> {
    const driver = this.context.primaryDriver;

    this.tool.debug(
      `Executing command ${chalk.magenta(driver.metadata.bin)} with args "${args.join(' ')}"`,
    );

    this.tool.emit('before-execute', [driver, args, this.context]);

    return this.executeCommand(driver.metadata.bin, args, { env: driver.options.env })
      .then(response => {
        driver.handleSuccess(response);

        this.tool.emit('after-execute', [driver, response]);

        return response;
      })
      .catch(error => {
        driver.handleFailure(error);

        this.tool.emit('failed-execute', [driver, error]);

        throw error;
      });
  }
}
