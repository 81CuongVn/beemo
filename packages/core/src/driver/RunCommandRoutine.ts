/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { Routine, Task } from '@boost/core';
import chalk from 'chalk';
import glob from 'fast-glob';
import path from 'path';
import fs from 'fs-extra';
import isGlob from 'is-glob';
import merge from 'lodash/merge';
import optimal, { array, bool, string } from 'optimal';
import parseArgs from 'yargs-parser';
import DriverContext from '../contexts/DriverContext';
import { STRATEGY_COPY } from '../constants';
import { Argv, Execution, BeemoTool } from '../types';

const OPTION_PATTERN: RegExp = /-?-[a-z0-9-]+(,|\s)/giu;

export type OptionMap = { [option: string]: true };

export interface RunCommandOptions {
  additionalArgv: Argv;
  forceConfigOption: boolean;
  workspaceRoot: string;
}

export default class RunCommandRoutine extends Routine<
  DriverContext,
  BeemoTool,
  RunCommandOptions
> {
  bootstrap() {
    this.options = optimal(
      this.options,
      {
        additionalArgv: array(string()),
        forceConfigOption: bool(),
        workspaceRoot: string().empty(),
      },
      {
        name: 'RunCommandRoutine',
      },
    );
  }

  async execute(context: DriverContext): Promise<Execution> {
    const { tool } = this;
    const { forceConfigOption, workspaceRoot } = this.options;
    const { metadata } = context.primaryDriver;

    this.task(tool.msg('app:driverRunGatherArgs'), this.gatherArgs);

    this.task(tool.msg('app:driverRunExpandGlob'), this.expandGlobPatterns);

    this.task(tool.msg('app:driverRunFilterOptions'), this.filterUnknownOptions).skip(
      !metadata.filterOptions,
    );

    if (workspaceRoot && metadata.workspaceStrategy === STRATEGY_COPY) {
      this.task(tool.msg('app:driverRunCopyWorkspaceConfig'), this.copyConfigToWorkspace);
    } else {
      this.task(tool.msg('app:driverRunIncludeConfigOption'), this.includeConfigOption).skip(
        !metadata.useConfigOption && !forceConfigOption,
      );
    }

    this.task(tool.msg('app:driverRunCommand'), this.runCommandWithArgs);

    return this.serializeTasks();
  }

  /**
   * When workspaces are enabled, some drivers require the config to be within each workspace,
   * instead of being referenced from the root, so we need to copy it.
   */
  async copyConfigToWorkspace(context: DriverContext, argv: Argv): Promise<Argv> {
    const { workspaceRoot } = this.options;

    this.debug('Copying config files to workspace');

    context.configPaths.forEach(config => {
      fs.copyFileSync(config.path, path.join(workspaceRoot, path.basename(config.path)));
    });

    return argv;
  }

  /**
   * Expand arguments that look like globs.
   */
  async expandGlobPatterns(context: DriverContext, argv: Argv): Promise<Argv> {
    const nextArgv: Argv = [];

    this.debug('Expanding glob patterns');

    argv.forEach(arg => {
      if (arg.charAt(0) !== '-' && isGlob(arg)) {
        const paths = glob
          .sync(arg, {
            cwd: context.root,
            onlyDirectories: false,
            onlyFiles: false,
          })
          .map(filePath => String(filePath));

        this.debug(
          '  %s %s %s',
          arg,
          chalk.gray('->'),
          paths.length > 0 ? paths.join(', ') : chalk.gray(this.tool.msg('app:noMatch')),
        );

        nextArgv.push(...paths);
      } else {
        nextArgv.push(arg);
      }
    });

    return nextArgv;
  }

  /**
   * Extract native supported options and flags from driver help output.
   */
  async extractNativeOptions(): Promise<OptionMap> {
    const driver = this.context.primaryDriver;
    const { env } = driver.options;
    const options = driver.getSupportedOptions();

    if (options.length > 0) {
      this.debug('Using supported options from driver');

      const nativeOptions: OptionMap = {};

      options.forEach(option => {
        nativeOptions[option] = true;
      });

      return Promise.resolve(nativeOptions);
    }

    this.debug('Extracting native options from help output');

    const { stdout } = await this.executeCommand(
      driver.metadata.bin,
      driver.metadata.helpOption.split(' '),
      {
        env,
      },
    );

    const nativeOptions: OptionMap = {};
    const matches = stdout.match(OPTION_PATTERN) || [];

    matches.forEach(option => {
      // Trim trailing comma or space
      nativeOptions[option.slice(0, -1)] = true;
    });

    return nativeOptions;
  }

  /**
   * Filter unknown and or unsupported CLI options from the arguments passed to the CLI.
   * Utilize the driver's help option/command to determine accurate options.
   */
  async filterUnknownOptions(context: DriverContext, argv: Argv): Promise<Argv> {
    this.debug('Filtering unknown command line options');

    const nativeOptions = await this.extractNativeOptions();
    const filteredArgv: Argv = [];
    const unknownArgv: Argv = [];
    let skipNext = false;

    argv.forEach((arg, i) => {
      if (skipNext) {
        skipNext = false;

        return;
      }

      if (arg.startsWith('-')) {
        let option = arg;
        const nextArg = argv[i + 1];

        // --opt=123
        if (option.includes('=')) {
          [option] = option.split('=');

          if (!nativeOptions[option]) {
            unknownArgv.push(arg);

            return;
          }

          // --opt 123
        } else if (!nativeOptions[option]) {
          unknownArgv.push(arg);

          if (nextArg && !nextArg.startsWith('-')) {
            skipNext = true;
            unknownArgv.push(nextArg);
          }

          return;
        }
      }

      filteredArgv.push(arg);
    });

    if (unknownArgv.length > 0) {
      this.debug('Filtered args: %s', chalk.gray(unknownArgv.join(', ')));
    }

    return filteredArgv;
  }

  /**
   * Gather arguments from all sources to pass to the driver.
   */
  async gatherArgs(context: DriverContext): Promise<Argv> {
    this.debug('Gathering arguments to pass to driver');

    const argv = [
      // Passed by the driver
      ...this.getDriverArgs(),
      // Passed on the command line
      ...this.getCommandLineArgs(),
      // Passed with parallel "//" operator
      ...this.getAdditionalArgs(),
    ];

    // Since we combine multiple args, we need to rebuild this.
    // And we need to set this before we filter them.
    // And we need to be sure not to remove existing args.
    context.args = merge({}, parseArgs(argv), context.args);

    return argv;
  }

  /**
   * Run some validation on additional/parallel args.
   */
  getAdditionalArgs(): Argv {
    const argv = this.options.additionalArgv;

    this.debug.invariant(argv.length > 0, 'From parallel operator', argv.join(' '), 'No arguments');

    return argv;
  }

  /**
   * Return args from the command line.
   */
  getCommandLineArgs(): Argv {
    const { argv } = this.context;

    this.debug.invariant(argv.length > 0, 'From the command line', argv.join(' '), 'No arguments');

    return argv;
  }

  /**
   * Return args from the primary driver.
   */
  getDriverArgs(): Argv {
    const argv = this.context.primaryDriver.getArgs();

    this.debug.invariant(
      argv.length > 0,
      'From driver "args" option',
      argv.join(' '),
      'No arguments',
    );

    return argv;
  }

  /**
   * Include --config option if driver requires it (instead of auto-lookup resolution).
   */
  async includeConfigOption(context: DriverContext, prevArgv: Argv): Promise<Argv> {
    const { primaryDriver } = context;
    const configPath = context.findConfigByName(primaryDriver.metadata.configName);
    const argv = [...prevArgv];

    if (configPath && primaryDriver.metadata.configOption) {
      argv.push(primaryDriver.metadata.configOption, configPath.path);
    }

    this.debug('Including config option to args');

    return argv;
  }

  /**
   * Execute the driver's command with the filtered arguments and handle the
   * success and failures with the driver itself.
   */
  async runCommandWithArgs(
    context: DriverContext,
    argv: Argv,
    task: Task<DriverContext>,
  ): Promise<Execution> {
    const driver = context.primaryDriver;
    const cwd = this.options.workspaceRoot || context.root;
    let result = null;

    this.debug(
      'Executing command "%s %s" in %s',
      chalk.magenta(driver.metadata.bin),
      argv.join(' '),
      chalk.cyan(cwd),
    );

    this.tool.emit(`${driver.name}.before-execute`, [context, argv, driver]);

    try {
      result = await this.executeCommand(driver.metadata.bin, argv, {
        cwd,
        env: driver.options.env,
        task,
      });

      driver.handleSuccess(result);

      this.tool.emit(`${driver.name}.after-execute`, [context, result, driver]);
    } catch (error) {
      driver.handleFailure(error);

      this.tool.emit(`${driver.name}.failed-execute`, [context, error, driver]);

      throw error;
    }

    return result;
  }
}
