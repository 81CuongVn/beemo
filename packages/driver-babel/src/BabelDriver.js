/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import path from 'path';
import rimraf from 'rimraf';
import { Driver } from '@beemo/core';

// Success: Writes file list to stdout
// Failure: Throws SyntaxError to stderr
export default class BabelDriver extends Driver {
  bootstrap() {
    this.setMetadata({
      bin: 'babel',
      configName: '.babelrc',
      description: 'Transpile files with Babel.',
      title: 'Babel',
    });

    this.setCommandOptions({
      clean: {
        boolean: true,
        default: true,
        description: 'Clean the target folder',
      },
    });

    this.on('babel.before-execute', this.handleCleanTarget);
  }

  /**
   * Automatically clean the target folder if --out-dir is used.
   */
  handleCleanTarget = (event: Event, driver: Driver, args: string[], context: Object) => {
    const options = context.yargs;

    if (options.clean && options.outDir) {
      rimraf.sync(path.resolve(options.outDir));
    }
  };
}
