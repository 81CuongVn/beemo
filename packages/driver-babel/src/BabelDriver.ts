/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import path from 'path';
import rimraf from 'rimraf';
import { Driver, DriverContext } from '@beemo/core';

// Success: Writes file list to stdout
// Failure: Throws SyntaxError to stderr
export default class BabelDriver extends Driver {
  bootstrap() {
    this.setMetadata({
      bin: 'babel',
      configName: '.babelrc',
      description: 'Transpile files with Babel.',
      title: 'Babel',
    });

    this.setCommandOptions({
      clean: {
        boolean: true,
        default: true,
        description: 'Clean the target folder',
      },
    });

    this.on('babel.before-execute', this.handleCleanTarget);
  }

  /**
   * Automatically clean the target folder if --out-dir is used.
   */
  handleCleanTarget = (driver: Driver, argList: string[], context: DriverContext) => {
    const { args } = context;

    if (args.clean && args.outDir) {
      rimraf.sync(path.resolve(args.outDir));
    }
  };
}
