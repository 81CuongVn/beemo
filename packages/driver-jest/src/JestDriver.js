/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import { Driver } from '@beemo/core';

import type { Execution } from '@beemo/core';

// Success: Writes passed tests to stderr (Bug? https://github.com/facebook/jest/issues/5064)
// Success: Writes coverage to stdout
// Failure: Writes failed tests to stderr
export default class JestDriver extends Driver {
  bootstrap() {
    this.setMetadata({
      bin: 'jest',
      configName: 'jest.json',
      dependencies: ['babel'],
      description: 'Unit test files with Jest.',
      filterOptions: true,
      title: 'Jest',
    });
  }

  // https://github.com/nodejs/node/issues/19218
  getSupportedOptions(): string[] {
    return [
      '--all',
      '--automock',
      '-b',
      '--bail',
      '--browser',
      '-c',
      '--cache',
      '--cacheDirectory',
      '--changedFilesWithAncestor',
      '--changedSince',
      '--ci',
      '--clearCache',
      '--clearMocks',
      '--collectCoverage',
      '--collectCoverageFrom',
      '--collectCoverageOnlyFrom',
      '--color',
      '--colors',
      '--config',
      '--coverage',
      '--coverageDirectory',
      '--coveragePathIgnorePatterns',
      '--coverageReporters',
      '--coverageThreshold',
      '--debug',
      '--detectLeaks',
      '-e',
      '--env',
      '--expand',
      '-f',
      '--findRelatedTests',
      '--forceExit',
      '--globalSetup',
      '--globalTeardown',
      '--globals',
      '--haste',
      '-h',
      '--help',
      '-i',
      '--json',
      '--lastCommit',
      '--listTests',
      '--logHeapUsage',
      '--mapCoverage',
      '--maxWorkers',
      '-w',
      '--moduleDirectories',
      '--moduleFileExtensions',
      '--moduleNameMapper',
      '--modulePathIgnorePatterns',
      '--modulePaths',
      '--noStackTrace',
      '--notify',
      '--notifyMode',
      '-o',
      '--onlyChanged',
      '--onlyFailures',
      '--outputFile',
      '--passWithNoTests',
      '--preset',
      '--projects',
      '--reporters',
      '--resetMocks',
      '--resetModules',
      '--resolver',
      '--restoreMocks',
      '--rootDir',
      '--roots',
      '--runInBand',
      '--runTestsByPath',
      '--setupFiles',
      '--setupTestFrameworkScriptFile',
      '--showConfig',
      '--silent',
      '--snapshotSerializers',
      '-t',
      '--testEnvironment',
      '--testFailureExitCode',
      '--testLocationInResults',
      '--testMatch',
      '--testNamePattern',
      '--testPathIgnorePatterns',
      '--testPathPattern',
      '--testRegex',
      '--testResultsProcessor',
      '--testRunner',
      '--testURL',
      '--timers',
      '--transform',
      '--transformIgnorePatterns',
      '-u',
      '--unmockedModulePathPatterns',
      '--updateSnapshot',
      '--useStderr',
      '--verbose',
      '-w',
      '--watch',
      '--watchAll',
      '--watchPathIgnorePatterns',
      '--watchman',
    ];
  }

  handleSuccess(response: Execution) {
    const out = response.stdout.trim();
    const err = response.stderr.trim();

    if (response.cmd.includes('--coverage')) {
      if (err) {
        this.tool.log(err);
      }

      if (out) {
        this.tool.log(out);
      }
    } else if (err) {
      this.tool.log(err);
    }
  }
}
