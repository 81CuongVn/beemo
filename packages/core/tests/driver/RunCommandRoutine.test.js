import { Tool } from 'boost';
import fs from 'fs-extra';
import RunCommandRoutine from '../../src/driver/RunCommandRoutine';
import BabelDriver from '../../../driver-babel/src/BabelDriver';
import JestDriver from '../../../driver-jest/src/JestDriver';
import {
  createDriverContext,
  setupMockTool,
  prependRoot,
  getRoot,
} from '../../../../tests/helpers';

jest.mock('fs-extra');

jest.mock('boost/lib/Tool');

const BABEL_HELP = `
Usage: babel [options] <files ...>


Options:

  -f, --filename [filename]            filename to use when reading from stdin - this will be used in source-maps, errors etc
  --retain-lines                       retain line numbers - will result in really ugly code
  --no-highlight-code                  enable/disable ANSI syntax highlighting of code frames (on by default) (default: true)
  --presets [list]
  --plugins [list]
  --ignore [list]                      list of glob paths to **not** compile
  --only [list]                        list of glob paths to **only** compile
  --no-comments                        write comments to generated output (true by default) (default: true)
  --compact [booleanString]            do not include superfluous whitespace characters and line terminators [true|false|auto]
  --minified                           save as much bytes when printing [true|false]
  -s, --source-maps [booleanString]    [true|false|inline]
  --source-map-target [string]         set \`file\` on returned source map
  --source-file-name [string]          set \`sources[0]\` on returned source map
  --source-root [filename]             the root from which all sources are relative
  --no-babelrc                         Whether or not to look up .babelrc and .babelignore files (default: true)
  --source-type [string]
  --auxiliary-comment-before [string]  print a comment before any injected non-user code
  --auxiliary-comment-after [string]   print a comment after any injected non-user code
  --module-root [filename]             optional prefix for the AMD module formatter that will be prepend to the filename on module definitions
  -M, --module-ids                     insert an explicit id for modules
  --module-id [string]                 specify a custom name for module ids
  --parser-opts [string]               Options to pass into the parser, or to change parsers (parserOpts.parser)
  --generator-opts [string]            Options to pass into the generator, or to change generators (generatorOpts.generator)
  -x, --extensions [extensions]        List of extensions to compile when a directory has been input [.es6,.js,.es,.jsx]
  -w, --watch                          Recompile files on changes
  --skip-initial-build                 Do not compile files before watching
  -o, --out-file [out]                 Compile all input files into a single file
  -d, --out-dir [out]                  Compile an input directory of modules into an output directory
  -D, --copy-files                     When compiling a directory copy over non-compilable files
  -q, --quiet                          Don't log anything
  -V, --version                        output the version number
  -h, --help                           output usage information
`;

describe('RunCommandRoutine', () => {
  let routine;
  let driver;
  let tool;

  beforeEach(() => {
    tool = setupMockTool(new Tool());

    driver = new BabelDriver({
      args: ['--qux'],
      env: { DEV: true },
    });
    driver.tool = tool;
    driver.bootstrap();

    routine = new RunCommandRoutine('babel', 'Run babel');
    routine.context = createDriverContext(driver);
    routine.tool = tool;
    routine.debug = jest.fn();
    routine.debug.invariant = jest.fn();
  });

  describe('bootstrap()', () => {
    it('errors if `forceConfigOption` is not a boolean', () => {
      routine.options.forceConfigOption = 'foo';

      expect(() => {
        routine.bootstrap();
      }).toThrowError('Invalid RunCommandRoutine field "forceConfigOption". Must be a boolean.');
    });

    it('errors if `workspaceRoot` is not a string', () => {
      routine.options.workspaceRoot = 123;

      expect(() => {
        routine.bootstrap();
      }).toThrowError('Invalid RunCommandRoutine field "workspaceRoot". Must be a string.');
    });
  });

  describe('execute()', () => {
    beforeEach(() => {
      routine.executeCommand = jest.fn(() => Promise.resolve({ stdout: BABEL_HELP }));
      driver.metadata.filterOptions = true;
    });

    it('executes pipeline in order', async () => {
      routine.context.argv.push('--out-dir', './lib');

      const argSpy = jest.spyOn(routine, 'gatherArgs');
      const globSpy = jest.spyOn(routine, 'expandGlobPatterns');
      const filterSpy = jest.spyOn(routine, 'filterUnknownOptions');
      const optSpy = jest.spyOn(routine, 'includeConfigOption');
      const runSpy = jest.spyOn(routine, 'runCommandWithArgs');

      const response = await routine.execute(routine.context);

      expect(argSpy).toHaveBeenCalledWith(routine.context, undefined, expect.anything());
      expect(globSpy).toHaveBeenCalledWith(
        routine.context,
        ['--qux', '-a', '--foo', 'bar', 'baz', '--out-dir', './lib'],
        expect.anything(),
      );
      expect(filterSpy).toHaveBeenCalledWith(
        routine.context,
        ['--qux', '-a', '--foo', 'bar', 'baz', '--out-dir', './lib'],
        expect.anything(),
      );
      expect(optSpy).not.toHaveBeenCalled();
      expect(runSpy).toHaveBeenCalledWith(
        routine.context,
        ['baz', '--out-dir', './lib'],
        expect.anything(),
      );
      expect(response).toEqual({ stdout: BABEL_HELP });
    });

    it('includes config option if `useConfigOption` is true', async () => {
      driver.metadata.useConfigOption = true;
      routine.context.configPaths.push(prependRoot(driver.metadata.configName));

      const optSpy = jest.spyOn(routine, 'includeConfigOption');
      const runSpy = jest.spyOn(routine, 'runCommandWithArgs');

      await routine.execute(routine.context);

      expect(optSpy).toHaveBeenCalledWith(routine.context, ['baz'], expect.anything());
      expect(runSpy).toHaveBeenCalledWith(
        routine.context,
        ['baz', '--config', prependRoot(driver.metadata.configName)],
        expect.anything(),
      );
    });

    it('doesnt filter unknown if `filterOptions` is false', async () => {
      driver.metadata.filterOptions = false;

      const filterSpy = jest.spyOn(routine, 'filterUnknownOptions');

      await routine.execute(routine.context);

      expect(filterSpy).not.toHaveBeenCalled();
    });

    it('calls `copyConfigToWorkspace` when driver is workspaces enabled', async () => {
      driver.metadata.workspaceStrategy = 'copy';
      routine.options.workspaceRoot = '/some/root';

      const copySpy = jest.spyOn(routine, 'copyConfigToWorkspace');

      await routine.execute(routine.context);

      expect(copySpy).toHaveBeenCalledWith(routine.context, ['baz'], expect.anything());
    });

    it('doesnt call `copyConfigToWorkspace` when driver is not workspaces enabled', async () => {
      routine.options.workspaceRoot = '/some/root';

      const copySpy = jest.spyOn(routine, 'copyConfigToWorkspace');

      await routine.execute(routine.context);

      expect(copySpy).not.toHaveBeenCalled();
    });

    it('doesnt call `copyConfigToWorkspace` when no workspace root', async () => {
      driver.metadata.workspaceStrategy = 'copy';

      const copySpy = jest.spyOn(routine, 'copyConfigToWorkspace');

      await routine.execute(routine.context);

      expect(copySpy).not.toHaveBeenCalled();
    });
  });

  describe('copyConfigToWorkspace()', () => {
    it('copies each config into workspace root', async () => {
      routine.options.workspaceRoot = '/some/root';
      routine.context.configPaths = ['.babelrc', 'jest.json'];

      const args = await routine.copyConfigToWorkspace(routine.context, ['foo', '--bar']);

      expect(args).toEqual(['foo', '--bar']);
      expect(fs.copyFileSync).toHaveBeenCalledWith('.babelrc', '/some/root/.babelrc');
      expect(fs.copyFileSync).toHaveBeenCalledWith('jest.json', '/some/root/jest.json');
    });
  });

  describe('expandGlobPatterns()', () => {
    it('passes through if no globs', async () => {
      const args = await routine.expandGlobPatterns(routine.context, ['--foo', 'bar', '-z']);

      expect(args).toEqual(['--foo', 'bar', '-z']);
    });

    it('converts globs to paths', async () => {
      const args = await routine.expandGlobPatterns(routine.context, [
        '--foo',
        '../{scripts,tests}/*.{sh,js}',
        'bar',
      ]);

      expect(args).toEqual([
        '--foo',
        '../scripts/build-packages.sh',
        '../scripts/bump-peer-deps.js',
        '../scripts/link-packages.sh',
        '../tests/helpers.js',
        'bar',
      ]);
    });
  });

  describe('extractNativeOptions()', () => {
    it('extracts all types of options', async () => {
      const options = await routine.extractNativeOptions();

      expect(options).toEqual(
        expect.objectContaining({
          '-f': true, // Short
          '-M': true, // Uppercase short
          '--filename': true, // Long
          '--no-highlight-code': true, // With dashes
        }),
      );
    });

    it('supports uppercased options', async () => {
      driver = new JestDriver();
      driver.tool = routine.tool;
      driver.bootstrap();

      routine.context.primaryDriver = driver;

      const options = await routine.extractNativeOptions();

      expect(options).toEqual(
        expect.objectContaining({
          '-u': true, // Short
          '--all': true, // Long
          '--changedFilesWithAncestor': true, // Camel case
        }),
      );
    });
  });

  describe('filterUnknownOptions()', () => {
    beforeEach(() => {
      routine.executeCommand = jest.fn(() => Promise.resolve({ stdout: BABEL_HELP }));
    });

    it('returns supported options', async () => {
      const args = await routine.filterUnknownOptions(routine.context, [
        './src',
        '-o',
        './lib',
        '--out-dir',
        './dist',
        '--minified',
      ]);

      expect(args).toEqual(['./src', '-o', './lib', '--out-dir', './dist', '--minified']);
    });

    it('filters unsupported options', async () => {
      const args = await routine.filterUnknownOptions(routine.context, [
        './src',
        '--foo',
        '-o',
        './lib',
        '--out-dir',
        './dist',
        '-X',
        '--minified',
        '--bar',
      ]);

      expect(args).toEqual(['./src', '-o', './lib', '--out-dir', './dist', '--minified']);
      expect(routine.debug).toHaveBeenCalledWith('Filtered args: %s', '--foo, -X, --bar');
    });

    it('skips unsupported option setters', async () => {
      const args = await routine.filterUnknownOptions(routine.context, [
        '--foo',
        '123',
        '--bar=456',
        '-w',
        '-c',
        '789',
        '-c=666',
      ]);

      expect(args).toEqual(['-w']);
      expect(routine.debug).toHaveBeenCalledWith(
        'Filtered args: %s',
        '--foo, 123, --bar=456, -c, 789, -c=666',
      );
    });
  });

  describe('gatherArgs()', () => {
    it('merges driver and command line args', async () => {
      const args = await routine.gatherArgs(routine.context);

      expect(args).toEqual(['--qux', '-a', '--foo', 'bar', 'baz']);
    });

    it('rebuilds context yargs', async () => {
      expect(routine.context.args).toEqual(
        expect.objectContaining({
          _: ['baz'],
          a: true,
          foo: 'bar',
        }),
      );

      await routine.gatherArgs(routine.context);

      expect(routine.context.args).toEqual(
        expect.objectContaining({
          _: ['baz'],
          a: true,
          foo: 'bar',
          qux: true,
        }),
      );
    });
  });

  describe('includeConfigOption()', () => {
    it('does nothing if a config path doesnt match', async () => {
      const args = await routine.includeConfigOption(routine.context, ['--foo']);

      expect(args).toEqual(['--foo']);
    });

    it('appends config path for a match', async () => {
      routine.context.configPaths.push(prependRoot(driver.metadata.configName));

      const args = await routine.includeConfigOption(routine.context, ['--foo']);

      expect(args).toEqual(['--foo', '--config', prependRoot(driver.metadata.configName)]);
    });
  });

  describe('runCommandWithArgs()', () => {
    beforeEach(() => {
      routine.executeCommand = jest.fn(() => Promise.resolve({ success: true }));
      driver.handleSuccess = jest.fn();
      driver.handleFailure = jest.fn();
    });

    it('executes command with correct args', async () => {
      const task = {};

      await routine.runCommandWithArgs(routine.context, ['--wtf'], task);

      expect(routine.executeCommand).toHaveBeenCalledWith('babel', ['--wtf'], {
        cwd: getRoot(),
        env: { DEV: true },
        task,
      });
    });

    it('handles success using driver', async () => {
      const response = await routine.runCommandWithArgs(routine.context, ['--wtf']);

      expect(response).toEqual({ success: true });
      expect(driver.handleSuccess).toHaveBeenCalledWith({ success: true });
    });

    it('handles failure using driver', async () => {
      routine.executeCommand.mockImplementation(() => Promise.reject(new Error('Oops')));

      try {
        await routine.runCommandWithArgs(routine.context, ['--wtf']);
      } catch (error) {
        expect(driver.handleFailure).toHaveBeenCalledWith(error);
      }
    });

    it('triggers `before-execute` event', async () => {
      await routine.runCommandWithArgs(routine.context, ['--wtf']);

      expect(routine.tool.emit).toHaveBeenCalledWith('before-execute', [
        driver,
        ['--wtf'],
        routine.context,
      ]);
    });

    it('triggers `after-execute` event on success', async () => {
      await routine.runCommandWithArgs(routine.context, ['--wtf']);

      expect(routine.tool.emit).toHaveBeenCalledWith('after-execute', [driver, { success: true }]);
    });

    it('triggers `failed-execute` event on failure', async () => {
      routine.executeCommand.mockImplementation(() => Promise.reject(new Error('Oops')));

      try {
        await routine.runCommandWithArgs(routine.context, ['--wtf']);
      } catch (error) {
        expect(routine.tool.emit).toHaveBeenCalledWith('failed-execute', [driver, error]);
      }
    });
  });
});
