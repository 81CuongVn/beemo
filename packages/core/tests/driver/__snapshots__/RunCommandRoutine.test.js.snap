// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RunCommandRoutine getAdditionalArgs() errors if a double quoted value is passed 1`] = `"--parallel option does not support nested quoted values."`;

exports[`RunCommandRoutine getAdditionalArgs() errors if a single quoted value is passed 1`] = `"--parallel option does not support nested quoted values."`;
